installutils/helminstall/client.go

```diff
@@ -0,0 +1,205 @@
+package helminstall
+
+import (
+	"io/ioutil"
+	"os"
+
+	"github.com/spf13/afero"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/release"
+)
+
+//go:generate mockgen -destination mocks/mock_helm_client.go -source ./client.go
+
+const (
+	TempChartFilePermissions = os.FileMode(0644)
+	TempChartPrefix          = "temp-helm-chart"
+	helmNamespaceEnvVar      = "HELM_NAMESPACE"
+)
+
+// This interface implements the Helm CLI actions. The implementation relies on the Helm 3 libraries.
+type HelmClient interface {
+	// Prepare an installation object that can then be .Run() with a chart object
+	NewInstall(namespace, releaseName string, dryRun bool) (HelmInstall, *cli.EnvSettings, error)
+
+	// Prepare an un-installation object that can then be .Run() with a release name
+	NewUninstall(namespace string) (HelmUninstall, error)
+
+	// List the already-existing releases in the given namespace
+	ReleaseList(namespace string) (ReleaseListRunner, error)
+
+	// Returns the Helm chart archive located at the given URI (can be either an http(s) address or a file path)
+	DownloadChart(chartArchiveUri string) (*chart.Chart, error)
+
+	// Returns true if the release with the given name exists in the given namespace
+	ReleaseExists(namespace, releaseName string) (releaseExists bool, err error)
+}
+
+// an interface around Helm's action.Install struct
+type HelmInstall interface {
+	Run(chrt *chart.Chart, vals map[string]interface{}) (*release.Release, error)
+}
+
+// an interface around Helm's action.Uninstall struct
+type HelmUninstall interface {
+	Run(name string) (*release.UninstallReleaseResponse, error)
+}
+
+var _ HelmInstall = &action.Install{}
+var _ HelmUninstall = &action.Uninstall{}
+
+// interface around needed afero functions
+type TempFile interface {
+	NewTempFile(fs afero.Fs, dir, prefix string) (f afero.File, err error)
+	WriteFile(fs afero.Fs, filename string, data []byte, perm os.FileMode) error
```

Comment:
Can we put the `afero.Fs` on the struct instead so it's an implementation detail

Reply:
I also simplified the dependencies by combining `afero.fs` and `TempFile` into a single interface